src/NuGet.Clients/NuGet.PackageManagement.UI/Models/PackageDetailControlModel.cs

```diff
@@ -106,23 +112,30 @@ protected override Task CreateVersionsAsync(CancellationToken cancellationToken)
             // installVersion is null if the package is not installed
             var installedVersion = installedDependency?.VersionRange?.MinVersion;
 
-            var allVersions = _allPackageVersions?.OrderByDescending(v => v).ToArray();
-
-            // allVersions is null if server doesn't return any versions.
-            if (allVersions == null)
-            {
-                return Task.CompletedTask;
-            }
+            var allVersions = _allPackageVersions?.OrderByDescending(v => v.version.Version).ToList();
 
             // null, if no version constraint defined in package.config
-            var allowedVersions = _projectVersionConstraints.Select(e => e.VersionRange).FirstOrDefault() ?? VersionRange.All;
-            var allVersionsAllowed = allVersions.Where(v => allowedVersions.Satisfies(v.version)).ToArray();
-
+            VersionRange allowedVersions = _projectVersionConstraints.Select(e => e.VersionRange).FirstOrDefault();
             // null, if all versions are allowed to be install or update
-            var blockedVersions = allVersions
-                .Select(v => v.version)
-                .Where(v => !allVersionsAllowed.Any(allowed => allowed.version.Equals(v)))
-                .ToArray();
+            var blockedVersions = new List<NuGetVersion>(allVersions.Count);
+
+            List<(NuGetVersion version, bool isDeprecated)> allVersionsAllowed;
+            if (allowedVersions == null)
+            {
+                allowedVersions = VersionRange.All;
+                allVersionsAllowed = allVersions;
+            }
+            else
+            {
+                allVersionsAllowed = allVersions.Where(v => allowedVersions.Satisfies(v.version)).ToList();
+                foreach (var version in allVersions)
+                {
+                    if (!allVersionsAllowed.Any(a => a.version.Version.Equals(version.version.Version)))
```

Comment:
`version.version.Version` looks absurd.  Can the [first `version`](https://github.com/NuGet/NuGet.Client/pull/3833/files#diff-0b6aa73227964074dc06409994723b4f7d158fe4edc669b5c8229c1c3d235bd4R131) be more appropriately named?